### ISSUE TITLE
Implement Core Engine, Features, and Heuristic Bot (Workflow A Only)

### ISSUE THREAD

**Issue by athariandre** [athariandre/yasta-bot#1]
STRICT REQUIREMENTS — ZERO DEVIATION PERMITTED

This is Task 1 of the project.
Your output will be used as the foundation for RL training, so correctness, determinism, and fidelity to the spec are critical.
Nothing in this task may be skipped or reinterpreted.

You must implement exactly four modules:

env.py
features.py
heuristic_bot.py
utils.py

Use ONLY numpy, queue/deque, and base Python libraries.

Do NOT use: torch, TensorFlow, JAX, gym, pygame, or any external library.

This is NOT a full game.
This is the training environment API for a Reinforcement Learning system that will be built in Workflow B.

The RL training code depends on this environment being exact, fast, and deterministic.

✅ ABSOLUTE RULES FOR THIS TASK
✅ 1. You must follow the spec in /docs/project_spec.md

ONLY implement what is listed under Workflow A.

NEVER look ahead to Workflow B or C.

Do NOT introduce features or abstractions outside the Workflow A section.

Do NOT make design decisions beyond what is explicitly stated.

✅ 2. Determinism is MANDATORY

Every function must produce the same output given the same inputs.

No randomness allowed except via utils.set_seed()

No sampling, no jitter, no nondeterministic ordering

All BFS queues must process neighbors in deterministic coordinate order
(preferred: Up, Right, Down, Left)

✅ 3. Performance Constraints

Training will require thousands of steps per second, so:

Use numpy for grid operations

No recursion

No heavy Python loops if vectorization is possible

No unnecessary object creation

No copying large arrays unless required

BFS operations must be O(rows × cols)

All functions must avoid quadratic or cubic behavior.

✅ 4. No reinterpretation or “helpful adjustments”

You must implement EXACT behavior.
Do NOT modify rules to “improve” gameplay.
Do NOT add state resets, timers, score counters, or anything beyond spec.

✅ NOW IMPLEMENT THE FOLLOWING MODULES EXACTLY
✅ (A) env.py — Core Game Environment

Implement class:

class TronEnv:
    def __init__(self, rows=18, cols=20, seed=None)
    def reset(self, start_mode="mirror") -> obs
    def step(self, action_my: int, action_opp: int) -> (obs, reward_my, reward_opp, done, info)
    def legal_moves(self, who: str) -> List[int]
    def clone(self) -> "TronEnv"

✅ Environment Rules (STRICT)

Board: 18 rows × 20 columns

Coordinates: (row, col)

Two players: 'p1' and 'p2'

Spawn positions:

p1 at (rows//2, cols//4)

p2 at (rows//2, cols - cols//4)

Trails are permanent

Movement is SIMULTANEOUS

New head positions are computed BEFORE any collision logic

Crash if:

new head is outside board

new head enters any non-empty cell

Head-on collision (same cell): both crash → draw

Rewards returned here MUST be exactly:

reward_my = 0

reward_opp = 0

✅ info dict MUST contain:
{
  "p1_dead": bool,
  "p2_dead": bool,
  "turn": int
}

✅ Observation Tensor (REQUIRED FORMAT)

Return a numpy float32 array shaped (8, rows, cols):

my_trail

opp_trail

free

my_head

opp_head

my_last_dir (constant raster value across grid)

opp_last_dir

boosts (always zeros)

Order must match exactly.

✅ (B) features.py — Numpy-Optimized Game Analysis

Implement fast, deterministic utilities:

✅ 1. reachable_area(grid, head) -> int

BFS flood-fill

Count number of reachable empty cells

Deterministic queue order: Up, Right, Down, Left

✅ 2. voronoi_counts(grid, head_me, head_opp) -> (me, opp, tie)

Multi-source BFS:

Compute distances from both heads

Cells closer to me → me

Closer to opp → opp

Equal distance → tie

Must be O(rows*cols)

✅ 3. is_tunnel(grid, head) -> bool

Definition:
A position is considered a tunnel if:

It has ≤1 free neighbor; OR

It is entering a 1-wide corridor (implement short forward probe)

✅ 4. simulate_step(grid, head, action) -> (new_head, crashed)

DO NOT modify the input grid

Compute next head

Return:

new_head (tuple)

crashed=True/False

All these will be used inside heuristic evaluation and PPO training.

✅ (C) heuristic_bot.py — Deterministic Expert Policy

Implement:

def choose_action(env: TronEnv, me: str) -> int:

✅ Behavior (STRICT)

For each legal move:

Clone environment (deep copy)

Simulate my move

Opponent move guess = action that maximizes reachable_area(opponent)

Step cloned env once using both moves

Score =

area_me - area_opp
- tunnel_penalty
+ cut_bonus

Where:

tunnel_penalty = 5 if my new head is in a tunnel that reduces mobility

cut_bonus = 1 * cut_cells
(cells reachable by opponent before but not after move)

✅ Tie-breaking:

Greater Voronoi advantage (me_cells - opp_cells)

Greater next-turn mobility (legal move count next turn)

Deterministic fallback ordering by action ID (0 < 1 < 2 < 3)

MUST:

Use ONLY numpy and env/features

Avoid randomness

Be efficient (many calls per second)

This bot will be used for behavior cloning later — correctness matters.

✅ (D) utils.py

Implement:

✅ 1. set_seed(seed)

Seed Python’s random

Seed numpy’s RNG

✅ 2. Logger class

Minimal deterministic logger

Methods:

info(msg)

debug(msg)

No external logging frameworks

✅ (E) Additional Non-Negotiable Requirements
✅ Deep Copy Requirements

clone() must create a fully independent deep copy

No shared numpy arrays

No shared mutable Python structures

simulate_step MUST NOT modify its inputs

✅ Deterministic BFS

Always process neighbors in this order:

[(−1, 0), (0, +1), (+1, 0), (0, −1)]
# Up, Right, Down, Left

✅ Efficiency Requirements

Minimize Python loops

Use preallocated arrays if possible

Avoid Python recursion entirely

✅ (F) Deliverables

The agent must output exactly four files
